detekt-rules-coroutines/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunSwallowedCancellation.kt

```diff
@@ -169,67 +201,99 @@ class SuspendFunSwallowedCancellation(config: Config) :
+                    ?.successfulFunctionCallOrNull()
+                    ?.argumentMapping
+                    ?.get(elementArgument)
+                    ?.symbol
             }
 
-            else -> true
+            valueSymbol
+                ?.let {
+                    it.isCrossinline.not() && it.isNoinline.not()
+                }
+                ?: false
         }
+
+        else -> true
     }
 
-    @Suppress("ReturnCount")
-    private fun KtExpression.hasSuspendCalls(): Boolean {
-        return when (this) {
-            is KtForExpression -> {
-                val loopRangeIterator = bindingContext[BindingContext.LOOP_RANGE_ITERATOR_RESOLVED_CALL, loopRange]
-                val loopRangeHasNext =
-                    bindingContext[BindingContext.LOOP_RANGE_HAS_NEXT_RESOLVED_CALL, loopRange]
-                val loopRangeNext = bindingContext[BindingContext.LOOP_RANGE_NEXT_RESOLVED_CALL, loopRange]
-                listOf(loopRangeIterator, loopRangeHasNext, loopRangeNext).any {
-                    it?.resultingDescriptor?.isSuspend == true
-                }
-            }
+    private fun KtExpression.hasSuspendCalls(): Boolean = when (this) {
+        is KtForExpression -> {
+            val loopRangeReferences = analyze(this) {
+                mainReference?.resolveToSymbols()
+                    ?.filterIsInstance<KaNamedFunctionSymbol>()
+            }.orEmpty()
+            loopRangeReferences.any { it.isSuspend }
+        }
 
-            is KtCallExpression, is KtOperationExpression -> {
-                val resolvedCall = getResolvedCall(bindingContext) ?: return false
-                (resolvedCall.resultingDescriptor as? FunctionDescriptor)?.isSuspend == true
-            }
+        is KtCallExpression, is KtOperationExpression -> {
+            val symbol = analyze(this) {
+                resolveToCall()
+                    ?.successfulCallOrNull<KaCompoundVariableAccessCall>()
+                    ?.compoundOperation
+                    ?.operationPartiallyAppliedSymbol
+                    ?.signature
+                    ?.symbol
 
-            is KtNameReferenceExpression -> {
-                val resolvedCall = getResolvedCall(bindingContext) ?: return false
-                val propertyDescriptor = resolvedCall.resultingDescriptor as? PropertyDescriptor
-                propertyDescriptor?.fqNameSafe == COROUTINE_CONTEXT_FQ_NAME
+                    ?: resolveToCall()
+                        ?.successfulFunctionCallOrNull()
+                        ?.symbol as? KaNamedFunctionSymbol
             }
 
-            else -> {
-                false
+            symbol?.isSuspend == true
+        }
+
+        is KtNameReferenceExpression -> {
+            val symbolFqName = analyze(this) {
+                resolveToCall()
+                    ?.let {
+                        it.successfulFunctionCallOrNull()
+                            ?: it.successfulVariableAccessCall()
+                    }
```

Comment:
```suggestion
                resolveToCall()
                ?.successfulCallOrNull<KaCallableMemberCall<*, *>>()
```
`KaCallableMemberCall` is base class for method and access calls

Reply:
Done in https://github.com/detekt/detekt/pull/8512/commits/9c1e61e53f77cbafcb8a493ddd3d35d1f6b77922